### ISSUE TITLE
Arctic reads a lot of data during upserts - expected behavior?

### ISSUE THREAD

**Issue by FTuma** [bmoscon/cryptostore#149]
I've been running cryptostore with redis & arctic for some time now and I noticed that the execution times of the Arctic write operations seem to slowly increase with the amount of data stored in each collection. 

Looking at the mongodb slow query logs revealed that the bottleneck are reads which seem to go through a lot (~500-35000) of documents/keys before returning only one document even if it is using some index. This translates to **reading up to multiple GB for each write operation** which takes a few seconds even on a fast SSD.

I also tried to create another index {"sy":1, "s":1,"sg":1,"e":1 } to create a covered query including the sort operation but since it is still using the old index after that, it didn't work but maybe I made an mistake since I'm not an expert for mongodb. If I force mongodb to use the new index  via $hint, it seem to slightly improve the performance but I don't know how I could incorporate the $hint logic into arctic.

My question are:

_1. Is that behavior normal for Arctic and to be expected over time?_
_2. Is there any way to improve the write performance by any way (additional indices, more specific queries, smaller chunks)?_
_3. Do other backends (parquet & influx) show the same behavior over time?_

Here are 3 examples of slow queries for Bitfinex, Coinbase with storage sizes: 14GB, 22GB and 5GB

{"t":{"$date":"2021-01-17T04:49:51.853+01:00"},"s":"I",  "c":"COMMAND",  "id":51803,   "ctx":"conn1358","msg":"Slow query","attr":{"type":"command","ns":"arctic.BITFINEX","command":{"find":"BITFINEX","filter":{"sy":{"$in":["l3_book-ETH-USD"]},"$and":[{"s":{"$lte":{"$date":"2021-01-17T03:44:59.999Z"}}},{"e":{"$gte":{"$date":"2021-01-17T03:44:00.000Z"}}}]},"sort":{"sy":1,"s":1,"sg":1},"lsid":{"id":{"$uuid":"1af7328c-eb0c-4adb-84b2-bbbd5eda1d3f"}},"$db":"arctic","$readPreference":{"mode":"primaryPreferred"}},"planSummary":"IXSCAN { sy: 1, s: 1, sg: 1 }"**,"keysExamined":33565,"docsExamined":33565,"cursorExhausted":true,"numYields":1278,"nreturned":1,**"queryHash":"312DBCC1","planCacheKey":"498C4CDD","reslen":510366,"locks":{"ReplicationStateTransition":{"acquireCount":{"w":1279}},"Global":{"acquireCount":{"r":1279}},"Database":{"acquireCount":{"r":1279}},"Collection":{"acquireCount":{"r":1279}},"Mutex":{"acquireCount":{"r":1}}},"storage":{"data":{**"bytesRead":9264478462,"timeReadingMicros":23164804**}},"protocol":"op_msg",**"durationMillis":24359**}}

{"t":{"$date":"2021-01-17T04:52:09.496+01:00"},"s":"I",  "c":"COMMAND",  "id":51803,   "ctx":"conn1360","msg":"Slow query","attr":{"type":"command","ns":"arctic.COINBASE","command":{"find":"COINBASE","filter":{"sy":{"$in":["l2_book-BTC-USD"]},"$and":[{"s":{"$lte":{"$date":"2021-01-17T03:46:59.999Z"}}},{"e":{"$gte":{"$date":"2021-01-17T03:46:00.000Z"}}}]},"sort":{"sy":1,"s":1,"sg":1},"lsid":{"id":{"$uuid":"5c812260-1c52-48db-8406-72719dd34e12"}},"$db":"arctic","$readPreference":{"mode":"primaryPreferred"}},"planSummary":**"IXSCAN { sy: 1, s: 1, sg: 1 }"**,**"keysExamined":21539,"docsExamined":21539,"cursorExhausted":true,**"numYields":1391,"nreturned":1,"queryHash":"312DBCC1","planCacheKey":"498C4CDD","reslen":748069,"locks":{"ReplicationStateTransition":{"acquireCount":{"w":1392}},"Global":{"acquireCount":{"r":1392}},"Database":{"acquireCount":{"r":1392}},"Collection":{"acquireCount":{"r":1392}},"Mutex":{"acquireCount":{"r":1}}},"storage":{"data":{**"bytesRead":8763190820,"timeReadingMicros":25867414**}},"protocol":"op_msg",**"durationMillis":27086**}}

{"t":{"$date":"2021-01-17T04:51:31.800+01:00"},"s":"I",  "c":"COMMAND",  "id":51803,   "ctx":"conn1360","msg":"Slow query","attr":{"type":"command","ns":"arctic.DERIBIT","command":{"find":"DERIBIT","filter":{"sy":{"$in":["funding-BTC-PERPETUAL"]},"$and":[{"s":{"$lte":{"$date":"2021-01-17T23:59:59.999Z"}}},{"e":{"$gte":{"$date":"2021-01-17T00:00:00.000Z"}}}]},"sort":{"sy":1,"s":1,"sg":1},"lsid":{"id":{"$uuid":"5c812260-1c52-48db-8406-72719dd34e12"}},"$db":"arctic","$readPreference":{"mode":"primaryPreferred"}},"planSummary":"IXSCAN { sy: 1, s: 1, sg: 1 }",**"keysExamined":50,"docsExamined":50,"cursorExhausted":true**,"numYields":34,"nreturned":1,"queryHash":"312DBCC1","planCacheKey":"498C4CDD","reslen":2610257,"locks":{"ReplicationStateTransition":{"acquireCount":{"w":35}},"Global":{"acquireCount":{"r":35}},"Database":{"acquireCount":{"r":35}},"Collection":{"acquireCount":{"r":35}},"Mutex":{"acquireCount":{"r":1}}},"storage":{"data":{**"bytesRead":460281572,"timeReadingMicros":1201135**}},"protocol":"op_msg",**"durationMillis":1211**}}

**Comment by yohplala** [comment]
Hi @FTuma 
I cannot give any hint for Arctic because I am not using it.
For Parquet, cryptostore simply dumps new data in a new parquet file. There is no query made on already written data.
Parquet files are sorted by embedding in their name the timestamp at which file is witten.
Bests,

**Comment by bmoscon** [comment]
yeah book data cant be left at that chunksize long term unless you have a powerful mongo setup. I'd recommend changing the chunksize to hourly, but your redis needs will increase. Alternatively, you can periodically move data from one collection to another (which a different chunk size) or copy it to other storage (parquet, csv, etc).

**Comment by FTuma** [comment]
Thanks for your suggestions!

@bmoscon I checked out the trades data which already uses hourly chunks and it basically would have the same issue over the long-term. The problem would just be less severe for the moment because there are not that many documents to scan through but that number is also increasing together over time. 
Currently on average ~950 docs are scanned in ~650ms vs ~500 docs in ~200ms about 3 weeks earlier while Cryptostore was constantly running and collecting data in between. 
So, at least for me, that's not a sustainable long-term solution and I decided to switch to InfluxDB & Parquet which seem to have a constant performance over time.
